content/200-concepts/100-components/02-prisma-client/260-client-extensions/120-query.mdx

```diff
@@ -1,16 +1,168 @@
 ---
 title: 'Create custom Prisma Client queries'
-metaTitle: 'Prisma Client extensions (Preview)'
+metaTitle: 'Prisma Client extensions: query component (Preview)'
 metaDescription: 'Extend the functionality of Prisma Client'
 tocDepth: 3
 ---
 
 <TopBlock>
 
-<Admonition type="warning">
+From version 4.7.0, you can use the `query` [Prisma Client extensions](/concepts/components/prisma-client/client-extensions) component type to hook into the query life-cycle and modify an incoming query or its result.
 
-We plan to add `query` extensions to a future version of Prisma. This page is a placeholder.
-
-</Admonition>
+You can use the `query` extension component as an alternative to [middlewares](/concepts/components/prisma-client/middleware#performance-and-appropriate-use-cases). Unlike middlewares, the `query` extension component gives you [user isolation](/concepts/components/prisma-client/client-extensions#extended-clients), such as in a row-level security (RLS) scenario, and end-to-end type safety. [Learn more](#query-extensions-versus-middlewares).
 
 </TopBlock>
+
+## Enable the preview feature
+
+Before you create Prisma Client extensions, you must enable the `clientExtensions` feature flag in the `generator` block of your `schema.prisma` file, as follows:
+
+```prisma
+generator client {
+  provider        = "prisma-client-js"
+  previewFeatures = ["clientExtensions"]
+}
+```
+
+## Create custom queries
+
+Use the `$extends` [client-level method](/reference/api-reference/prisma-client-reference#client-methods) to create an [extended client](/concepts/components/prisma-client/client-extensions#about-prisma-client-extensions). An extended client is a variant of the standard Prisma Client that is wrapped by one or more extensions.
+
+Use the `query` extension component to create custom queries. You can add a custom query to the following:
+
+- [A specific operation in a specific model of your schema](#add-a-custom-query-to-a-specific-operation-in-a-specific-model)
+- [A specific operation in all models of your schema](#add-a-custom-query-to-a-specific-operation-in-all-models-of-your-schema)
+- [All operations in a specific model of your schema](#add-a-custom-query-to-all-operations-in-a-specific-model)
+- [All operations in all models of your schema](#add-a-custom-query-to-all-operations-in-all-models-of-your-schema)
+
+To create a custom query, use the following structure:
+
+```ts
+const xprisma = prisma.$extends({
+  name?: 'name',
+  query?: {
+    user: { ... } //in this case, we add a query to the `user` model
+  },
+});
```

Comment:
We could split this a bit further, to make clear what is the component and then what is used to warp it around the client:
```suggestion
const queryExtensionComponent = {
  name?: 'name',
  query?: {
    user: { ... } // in this case, we add a query to the `user` model
  },
}
const xprisma = prisma.$extends(queryExtensionComponent);
```
Is that valid and helpful @millsp?

Reply:
That is valid in JS, invalid is TS since you wouldn't benefit from type inference in your object. That's when you need to use `Prisma.defineExtension`.

Reply:
OK, I'll leave this as-is for now. We can refine at a later date.